WIP: Electron startup performance measurement script

cdamus opened the pull request:
Refactor a reusable module out of the startup measurement script. Use this to build a new script measuring the start-up performance of the Electron example. That script uses the Chrome tracing facility, initialized on start-up, to determine how long it takes until the loading indicator is hidden. This is inferred from the last occurrence in the trace of a `HitTest` event on the `DIV` element that has the `theia-preload` and `theia-hidden` classes.

Contributed on behalf of STMicroelectronics.

sdirix commented on package.json:
```diff
@@ -79,7 +79,8 @@
     "watch": "concurrently --kill-others -n tsc,browser,electron -c red,yellow,blue \"tsc -b -w --preserveWatchOutput\" \"yarn --cwd examples/browser watch:bundle\" \"yarn --cwd examples/electron watch:bundle\"",
     "watch:compile": "concurrently --kill-others -n cleanup,tsc -c magenta,red \"ts-clean dev-packages/* packages/* -w\" \"tsc -b -w --preserveWatchOutput\"",
-    "performance:startup": "concurrently --success first -k -r \"cd scripts/performance && node measure-performance.js --name Startup --folder startup --runs 10\" \"yarn --cwd examples/browser start\""
+    "performance:startup": "concurrently --success first -k -r \"cd scripts/performance && node measure-performance.js --name Startup --folder startup --runs 10\" \"yarn --cwd examples/browser start\"",
+    "performance:electron": "yarn electron rebuild && cd scripts/performance && node electron-performance.js"
```
* I think the scripts should either be `performance:browser` & `performance:electron` or `performance:startup:browser` & `performance:startup:electron`
 * If we rebuild for electron in the script then we should also rebuild for browser

cdamus replied:
Yeah, I didn't really want to rename an existing script because it's probably used by somebody's automation somewhere? I could alias it, but there's no deprecation mechanism for scripts, right?

I'll remove the rebuild. I hadn't intended it to remain.

cdamus replied:
Actually, it turns out that the Electron rebuild is needed because `yarn build:examples` doesn't do that step. And it's included indirectly in the browser start, so I don't need to add it to that script in the `package.json`.

sdirix commented on scripts/performance/electron-performance.js:
```diff
@@ -0,0 +1,99 @@
+
+    measure(name, condition, runs, testScenario, hasNonzeroTimestamp, isPreloadHidden);
+}
+
+async function launchElectron(traceConfigFile) {
```
We should annotate of what type `traceConfigFile` is. It's a bit confusing as there is also the global `traceConfigFile` so maybe we want to rename the parameter?

sdirix commented on scripts/performance/electron-performance.js:
```diff
@@ -0,0 +1,99 @@
+let traceFile = resolve(electronExample, 'electron-trace.json');
+let runs = 10;
+
+const traceConfigFile = resolve(__dirname, './electron-trace-config.json');
```
Maybe rename to `traceConfigPath`?

sdirix commented on scripts/performance/electron-performance.js:
```diff
@@ -0,0 +1,99 @@
+    process.on('SIGTERM', exitHandler(true));
+
+    /** @type import('./common-performance').TestFunction */
+    const testScenario = async (runNr) => {
```
`runNr` is unused?

cdamus replied:
> Does this only test the frontend part of Electron or does it also include the backend time?

The performance measurements start as soon as Electron is up. So as long as it's the Electron app that starts the back-end, then it should encompass both the front- and the back-end.

> For me it takes a long time but I don't think it's part of the performance script:
> 
> > [Performance][ElectronStartup][MEAN] UI Ready: 6.552 seconds
> > [Performance][ElectronStartup][STDEV] UI Ready: 0.262 seconds
> > Done in 152.87s.
> 
> Interestingly the browser performance step reports normal times. Starting the Electron example app manually also seems to take quite a while.

This is strange. I get pretty much exactly the same run-times as the browser script.

> * Can Electron be started headless? That would be much better than starting it up every time.

Will check.

> * The performance script crashes when the Electron app was not built before. Can we do a check whether the app exists and exit with an explanatory error message?

Should probably do the same in the browser script.

sdirix replied:
> The performance measurements start as soon as Electron is up. So as long as it's the Electron app that starts the back-end, then it should encompass both the front- and the back-end.

Can you verify? Maybe add a timeout in Electron before the frontend window is even opened.

cdamus replied:
With a delay in the `electron-main.js` before opening the window, I see that the test execution is delayed but the measurement is the same as before.

With a delay in the `main.js` of the back-end before calling the node `main` function, I see that again the test execution is delayed and again the measurement is the same as before.

So, it seems that we are measuring only the front-end. But that is the same in the browser case. The trace log starts only when the Electron UI is created, which it seems is after the back-end has been started. What we would need to include the back-end in the measurement is to grab a timestamp in the performance script when launching the app, then correlate the timestamps in the trace file to that baseline. I'm looking into it.

sdirix reviewed (changes requested):
Does this only test the frontend part of Electron or does it also include the backend time?

For me it takes a long time but I don't think it's part of the performance script:

>[Performance][ElectronStartup][1] UI Ready: 7.261 seconds                                                                                                                                                          
[Performance][ElectronStartup][2] UI Ready: 6.145 seconds                                                                                                                                                          
[Performance][ElectronStartup][3] UI Ready: 6.510 seconds                                                                                                                                                          
[Performance][ElectronStartup][4] UI Ready: 6.510 seconds                                                                                                                                                          
[Performance][ElectronStartup][5] UI Ready: 6.530 seconds                                                                                                                                                          
[Performance][ElectronStartup][6] UI Ready: 6.479 seconds                                                                                                                                                          
[Performance][ElectronStartup][7] UI Ready: 6.511 seconds                                                                                                                                                          
[Performance][ElectronStartup][8] UI Ready: 6.538 seconds                                                                                                                                                          
[Performance][ElectronStartup][9] UI Ready: 6.558 seconds                                                                                                                                                          
[Performance][ElectronStartup][10] UI Ready: 6.479 seconds                                                                                                                                                         
[Performance][ElectronStartup][MEAN] UI Ready: 6.552 seconds                                                                                                                                                       
[Performance][ElectronStartup][STDEV] UI Ready: 0.262 seconds                                                                                                                                                      
Done in 152.87s.

Interestingly the browser performance step reports normal times.
Starting the Electron example app manually also seems to take quite a while.

---

 * Can Electron be started headless? That would be much better than starting it up every time.
 * The performance script crashes when the Electron app was not built before. Can we do a check whether the app exists and exit with an explanatory error message?

cdamus commented:
Commit bb034f73538e74e557e7406e6dc056663ae6628e addresses review comments.

I can confirm that only front-end run-time is measured by the script as the tracing only because when Electron creates the Chromium browser. So, this measures essentially the same thing as the browser start-up performance script. Trying to include the back-end time in the measurement is complicated by the difficulty of determining when exactly the Chromium environment is initialized and the lack of access in the node environment to whatever monotonic clock Chromium is using for its measurements, which is not relative either to system start-up (as in the case of node's process.hrtime()) nor to process start-up. So including the back-end timing is left out of scope of this PR.

sdirix commented on scripts/performance/electron-performance.js:
```diff
@@ -0,0 +1,119 @@
+const electronExample = resolve(basePath, 'examples/electron');
+const theia = resolve(electronExample, 'node_modules/.bin/theia');
+
+let name = 'ElectronStartup';
```
To be more precise we should call this `ElectronFrontendStartup`

sdirix replied:
The other script should then also be adapted to `BrowserFrontendStartup`. Maybe also change its file name so the Browser vs Electron cases are clear?

sdirix reviewed (changes requested):
During the end of my first run I got this error message:

>[Performance][ElectronStartup][8] UI Ready: 2.400 seconds
[Performance][ElectronStartup][9] UI Ready: 2.422 seconds
(node:2156791) UnhandledPromiseRejectionWarning: Error: Could not analyze performance trace
    at analyzeTrace (/home/stefan/Git/theia/scripts/performance/common-performance.js:110:11)
    at measure (/home/stefan/Git/theia/scripts/performance/common-performance.js:62:28)
(node:2156791) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:2156791) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Done in 153.18s.

My second test run immediately failed on the second iteration:

>[Performance][ElectronStartup][1] UI Ready: 2.385 seconds
(node:2162318) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'startsWith' of undefined
    at isPreloadHidden (/home/stefan/Git/theia/scripts/performance/electron-performance.js:118:45)
    at /home/stefan/Git/theia/scripts/performance/common-performance.js:103:16
    at Array.filter (<anonymous>)
    at analyzeTrace (/home/stefan/Git/theia/scripts/performance/common-performance.js:98:43)
    at measure (/home/stefan/Git/theia/scripts/performance/common-performance.js:62:28)
(node:2162318) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not h
andled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejectio
n id: 2)
(node:2162318) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Done in 32.43s.

The third test run failed on the fourth iteration:

>(node:2163911) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'startsWith' of undefined
    at isPreloadHidden (/home/stefan/Git/theia/scripts/performance/electron-performance.js:118:45)

cdamus commented:
Commit 4b5a51569e83c1347f9cf92ea1baadeba931b9df renames scripts as suggested and adds some hardening measures to

- more gracefully handle failures to analyze the trace output, for whatever reason
- ensure that the workspace directory exists, in case maybe on some platform it matters
- generate an unique trace output per test run, to enable diagnosis of any run failures that may occur

This now more closely aligns trace file management with the browser performance script and should help to get cleaner test execution and, where necessary, enough information to diagnose what went wrong in test execution.

@sdirix I suspect that on your system it may be necessary to increase the duration of tracing in the `electron-trace-config.json` file. This currently specifies a 10 seconds duration. Can you try 15 and let me know whether that makes test execution more reliable? This may be something that needs to be externally parameterized for eventual CI in shared server environment.

cdamus commented:
@sdirix commit 24fc529151f14fdae33a15c5acd8b888e3636eca updates the Extension Impact script to use the new name of the browser performance script and makes changes in the performance scripts readme:

- updates for changes in names and default command-line arguments in the browser script
- document the Electron start-up script similarly to the browser script

sdirix commented on CHANGELOG.md:
```diff
@@ -1,5 +1,7 @@
 # Change Log
 
+- [scripts] added Electron front-end start-up performance measurement script [#](https://github.com/eclipse-theia/theia/pull/#) - Contributed on behalf of STMicroelectronics
```
```suggestion
- [scripts] added Electron frontend start-up performance measurement script [#](https://github.com/eclipse-theia/theia/pull/#) - Contributed on behalf of STMicroelectronics
```

sdirix commented on scripts/performance/README.md:
```diff
@@ -1,13 +1,16 @@
 
-This directory contains a script that measures the performance of Theia.
-Currently the support is limited to measuring the `browser-app`'s startup time using the `Largest contentful paint (LCP)` value.
+This directory contains script that measure the start-up performance of the Theia front-end in both the browser and the Electron examples.
```
```suggestion
This directory contains scripts that measure the start-up performance of the Theia frontend in both the browser and the Electron examples.
```

sdirix reviewed (commented):
The task now works without errors on my machine!

I noticed something interesting: When I clean the whole workspace and rebuild everything (i.e. `git clean -dfx && yarn && yarn build && yarn performance:startup:electron` the very first Electron run takes over a second longer. Do you know why that is and can you reproduce the issue? Subsequent runs always take roughly the same time.

sdirix commented:
> @sdirix I suspect that on your system it may be necessary to increase the duration of tracing in the `electron-trace-config.json` file. This currently specifies a 10 seconds duration. Can you try 15 and let me know whether that makes test execution more reliable? This may be something that needs to be externally parameterized for eventual CI in shared server environment.

Now it works without me doing any changes

cdamus commented:
Commit 720c7052a6e9ad0c6c6d5835a6a15e799b57c915 updates the Electron tracing to use the same LCP metric as the browser script, for consistency. However, it is unable to use the same trace event to find the start of tracing because that event is not emitted in the Electron implementation even when configured for the `disabled-by-default-devtools.timeline` category.

cdamus commented:
Commit 5999350151f4da337a7177ad27194e4d21fe9825 fixes the `extension-impact.js` script so that it will work when run without the `--base-time` option.

cdamus commented:
Commit bbfad6c126ed35b3e3cf5c078f424ba7b9ca7cdb squashes everything so far and also adds integration of the Electron start-up performance measurement script into the `extension-impact.js` script, which now can drive the tests in either the browser or Electron.

sdirix reviewed (approved):
LGTM!

sdirix commented:
Merged upstream with https://github.com/eclipse-theia/theia/pull/10442
